### PR TITLE
(on getgov-za) Ticket #2421/#2422:  Missing first name and last name gets a green checkmark next to “Unknown” and make field label consistent

### LINKED ISSUE #2421: User profile: missing first name and last name gets a green checkmark next to “Unknown”
### Current Behavior

Missing first name and last name gets a green checkmark next to “Unknown”.

image

### Expected Behavior

Any of the fields that are “missing” should show with a red exclamation mark.

### Steps to Reproduce

On scenario2: First time user profile set up ([as documented here](https://docs.google.com/document/d/1uzZy6a7u4uc9lLtfHfYpWQmGpmNFE_z7mE2GIDqhveY/edit))

- Log in to staging as a super admin, and navigate to /admin
- Go to the Users table
- Find your user record in the table, and edit Personal info and Contact info. Blank out one or more fields. Save.
- Navigate to  https://getgov-staging.app.cloud.gov/ home page. 
- You should see the the user profile with the headline: “Finish setting up your profile”.
- Any of the fields that are “missing” should show as “None” and with a red exclamation mark.

### Environment

_No response_

### Additional Context

_No response_

### Issue Links

_No response_

### LINKED ISSUE #2422: User profile:
### Issue description

When user updates multiple fields in their user profile information, the application will not save the updated information until all updated fields are valid. If one field is invalid, the field is highlighted, all other fields are collapsed and their labels show the last saved information. But, upon expanding the field, the newly entered information is displayed. There is an inconsistency in the label and the expanded views. This is confusing for the user.

### Acceptance criteria

_No response_

### Additional context

image

Expanded view:
image

### Links to other issues

_No response_

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #2421
Resolves #2422

## Changes

- When the input fields are changed, we update the "readonly label" accordingly such that these two fields are in sync.
- Minor code flow changes to account for this.

## Context for reviewers
A bug was found wherein the full name field on the Finish User Profile page would display the "complete" icon when we were missing data for the users name.

When the user would save in this state, the page would still show that these fields were missing data - but it was not immediately obvious as it displayed that it was complete.

This bug was being caused by an edge case regarding the "full_name" field being a) required and b) defaulted to "Unknown" on None instead of just being None.

This PR changes the flow such that full_name will always be none if we don't have any values for first name or last name. Then, we include javascript that updates this readonly value for the full name and other fields.

## Setup
1. Enable the profile flag
2. Edit your user such that you are missing some required data. More specifically, remove the name fields and an unrelated field like title.
3. Go to home (you should be redirected to the finish user setup page)
4. Edit one of your name fields and add a first name and last name. Do _not_ add a title. When you click save, your changes should now reflect on the name fields
5. Do the same for the name fields but in reverse such that your testing title.

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/forms/user_profile.py [hunk 1]
```diff
@@ -93,4 +93,10 @@ def __init__(self, *args, **kwargs):
...
 
         # Define the "full_name" value
-        self.fields["full_name"].initial = self.instance.get_formatted_name()
+        full_name = ""
+        if self.instance.first_name and self.instance.last_name:
+            full_name = self.instance.get_formatted_name(return_unknown_when_none=False)
+        self.fields["full_name"].initial = full_name
+
+        # Set full_name as required for styling purposes
+        self.fields["full_name"].widget.attrs['required'] = 'required'
```
Comment (rachidatecs) [inline src/registrar/forms/user_profile.py:102]: Why the workaround?

File: src/registrar/assets/js/get-gov.js [hunk 2]
```diff
@@ -1834,6 +1834,32 @@ document.addEventListener('DOMContentLoaded', function() {
...
+              let middleName = document.querySelector(`#id_middle_name`).value;
+              let lastName = document.querySelector(`#id_last_name`).value;
+              if (firstName && middleName && lastName) {
+                let values = [firstName.value, middleName.value, lastName.value]
+                readonlyField.innerHTML = values.join(" ");
+              }else {
+                readonlyField.innerHTML = "Unknown";
+              }
+
+              inputField.classList.add("text-base")
```
Comment (rachidatecs) [inline src/registrar/assets/js/get-gov.js:1859]: Why the JS class assignemnt? And where's the readonlyField.innerHTML = inputFieldValue (for phone and title)?

File: src/registrar/assets/js/get-gov.js [hunk 3]
```diff
@@ -1834,6 +1834,39 @@ document.addEventListener('DOMContentLoaded', function() {
...
+          if (!inputField || !readonlyField) {
+            return;
+          }
+
+          let inputFieldValue = inputField.value
+          if (inputFieldValue || fieldName == "full_name"){
+            if (fieldName == "full_name"){
+              let firstName = document.querySelector(`#id_first_name`);
+              let middleName = document.querySelector(`#id_middle_name`);
+              let lastName = document.querySelector(`#id_last_name`);
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/assets/js/get-gov.js:1851]: Agree, commit

File: src/registrar/tests/test_views.py [hunk 4]
```diff
@@ -577,6 +620,47 @@ def test_new_user_with_profile_feature_on(self):
             completed_setup_page = self.app.get(reverse("home"))
             self.assertContains(completed_setup_page, "Manage your domain")
 
+    @less_console_noise_decorator
+    def test_new_user_with_empty_name_profile_feature_on(self):
```
Comment (rachidatecs) [inline src/registrar/tests/test_views.py:624]: def test_new_user_with_empty_name_can_add_name(self):

File: src/registrar/assets/js/get-gov.js [hunk 5]
```diff
@@ -1834,6 +1834,39 @@ document.addEventListener('DOMContentLoaded', function() {
         
         // When the edit button is clicked, show the input field under it
         handleEditButtonClick(fieldName, button);
+
+        let editableFormGroup = button.parentElement.parentElement.parentElement;
```
Comment (CocoByte) [inline src/registrar/assets/js/get-gov.js:1838]: Seems using an ID or class would be better...is there an inconsistency in the DOM that prevents a more proper lookup?

File: src/registrar/assets/js/get-gov.js [hunk 6]
```diff
@@ -1834,6 +1834,39 @@ document.addEventListener('DOMContentLoaded', function() {
...
+              if (firstName && lastName) {
+                let values = [firstName.value, middleName.value, lastName.value]
+                readonlyField.innerHTML = values.join(" ");
+              }else {
+                readonlyField.innerHTML = "Unknown";
+              }
+              
+              // Technically, the full_name field is optional, but we want to display it as required. 
+              // This style is applied to readonly fields (gray text). This just removes it, as
+              // this is difficult to achieve otherwise by modifying the .readonly property.
```
Comment (CocoByte) [inline src/registrar/assets/js/get-gov.js:1861]: Appreciate the comments here

File: src/registrar/assets/sass/_theme/_base.scss [hunk 7]
```diff
@@ -190,7 +190,7 @@ abbr[title] {
     svg.usa-icon {
       color: #{$dhs-red};
     }
-    div.readonly-field {
+    div.input-with-edit-button__readonly-field {
```
Comment (CocoByte) [inline src/registrar/assets/sass/_theme/_base.scss:193]: This strips away the red coloring for other readonly fields.  Did we want to do this?

File: src/registrar/forms/user_profile.py [hunk 8]
```diff
@@ -93,4 +93,10 @@ def __init__(self, *args, **kwargs):
         self.fields["title"].label = "Title or role in your organization"
 
         # Define the "full_name" value
-        self.fields["full_name"].initial = self.instance.get_formatted_name()
+        full_name = ""
+        if self.instance.first_name and self.instance.last_name:
```
Comment (CocoByte) [inline src/registrar/forms/user_profile.py:97]: Seems like this logic should be inside of get_formatted_name(). What was the reasoning for doing these checks here?

File: src/registrar/models/contact.py [hunk 9]
```diff
@@ -87,10 +87,13 @@ def _has_more_than_one_join_per_relation(self, relation, expected_relation):
                 return getattr(self, relation).count() > threshold
         return False
 
-    def get_formatted_name(self):
+    def get_formatted_name(self, return_unknown_when_none=True):
```
Comment (CocoByte) [inline src/registrar/models/contact.py:90]: Is there ever a case where return_unknown_when_none != True?

**Review by CocoByte — CHANGES_REQUESTED** [review]
Got a few comments for you to look at.

**Review by kristinacyin — CHANGES_REQUESTED** [review]
After I deleted my first name and last name in admin, I saw this:

image

While I didn't get "Unknown", my answer was blank and I still got the green check mark

**Comment by kristinacyin** [comment]
Also, I'm going to crosspost this question in #dotgov-design for others' input:


I noticed that the order is slightly different between **/finish-profile-setup** and **/user-profile**

/user-profile has the order of Name, Title, Org Email, Phone
image

While /finish-profile-setup has the order of Name, Org Email, Title, Phone
image

I think we should make both of them consistent. Does it make more sense to have Org Email or Title first? Personally, it makes more sense to me to have Title first

**Review by kristinacyin — APPROVED** [review]
LGTM! 🥇

File: src/registrar/assets/js/get-gov.js (hunk 5 above)
Comment (CocoByte) [reply src/registrar/assets/js/get-gov.js:1841]: Seems like a red flag architecture-wise, but I see the reasoning.  Sounds good

File: src/registrar/assets/sass/_theme/_base.scss (hunk 7 above)
Comment (CocoByte) [reply src/registrar/assets/sass/_theme/_base.scss:193]: It's not really an issue...The only thing is that when we get specific like this we rely more and more on several "red" settings to basically stay in sync, which seems prone to breakages.  We do this alot already in our scss architecture, though, so I will leave it.  I've noticed that we are gradually refactoring our scss as we go so it shouldn't be too much of a problem.

File: src/registrar/forms/user_profile.py (hunk 8 above)
Comment (CocoByte) [reply src/registrar/forms/user_profile.py:97]: Got it...seems like an inconsistent UI rule, but I won't argue the ACs this time around.

**Review by CocoByte — APPROVED** [review]
LGTM!

**Pull request merged**